### PR TITLE
Add initial support and doc for ad2s1210

### PR DESCRIPTION
**Pull request by apelete**
This adds initial support and doc for AD2S1210 Resolver-to-Digital Converter.

Matlab implementation currently depends on AD2S1210 Linux driver work-in-progress implementation ([see AD2S1210 updates #2261](https://github.com/analogdevicesinc/linux/pull/2261)).

Consequently, this is an initial support proposal to build upon based on AD2S1210 Linux driver implementation.

### CODE REVIEW

File: +adi/+AD2S1210/Rx.m [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        function obj = Rx(varargin)
+            obj = obj@matlabshared.libiio.base(varargin{:});
+            obj.enableExplicitPolling = false;
+            obj.EnabledChannels = [1 2];
+            obj.BufferTypeConversionEnable = true;
+        end
+
+        %% Flush the buffer
+        function flush(obj)
+            flushBuffers(obj);
```
Comment (tfcollins) [inline +adi/+AD2S1210/Rx.m:83]: Need docstring

File: +adi/+AD2S1210/Rx.m [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+            if obj.ConnectedToDevice
+                obj.Velocity = obj.getAttributeRAW('anglvel0','raw', obj.isOutput);
+            end
+        end
+
+        %% Check Velocity Scale
+        function getVelocityScale(obj)
+            if obj.ConnectedToDevice
+                obj.VelocityScale = obj.getAttributeDouble('anglvel0', 'scale', obj.isOutput);
+            end
```
Comment (tfcollins) [inline +adi/+AD2S1210/Rx.m:111]: These all should be get.* methods to match convention.

**Review by tfcollins — CHANGES_REQUESTED** [review]
All the get* should be get.* methods. For the scale properties, are these writable? How are they used?

File: +adi/+AD2S1210/Rx.m (hunk 1 above)
Comment (apelete) [reply +adi/+AD2S1210/Rx.m:83]: Looked like a mandatory abstract method to be implemented but it does not seem to be the case, so I removed it.

**Comment by apelete** [comment]
> All the get* should be get.* methods. For the scale properties, are these writable? How are they used?

All renamed as asked.
Scale properties are read-only, these values are a multiplier to be applied to the corresponding "raw" property value to obtain the measured value in specified units.

File: +adi/+AD2S1210/Rx.m [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        end
+
+        %% Check Velocity Scale
+        function rValue = get.VelocityScale(obj)
+            if obj.ConnectedToDevice
+                rValue = obj.getAttributeDouble('anglvel0', 'scale', obj.isOutput);
+                obj.VelocityScale = rValue;
+            end
+        end
```
Comment (tfcollins) [inline +adi/+AD2S1210/Rx.m:111]: rValue is never returned if ConnectedToDevice is false. get methods should return NaN when not connected to hardware

File: +adi/+AD2S1210/Rx.m [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        Velocity = 0
+
+        % VelocityScale Velocity Angular Scale
+        %   Resolver velocityscale.
+        VelocityScale = 0
+
+        % ExcitationFrequency Excitation Frequency
+        %   Resolver excitation frequency in Hertz.
+        ExcitationFrequency = 0
+    end
```
Comment (tfcollins) [inline +adi/+AD2S1210/Rx.m:39]: Anything with a get. method should be a Dependent property https://www.mathworks.com/help/matlab/matlab_oop/access-methods-for-dependent-properties.html

File: +adi/+AD2S1210/Rx.m [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        end
+    end
+
+    %% API Functions
+    methods (Hidden, Access = protected)
+
+        function setupInit(obj)
+            % Read excitation frequency attribute from device once connected.
+
+            obj.ExcitationFrequency = obj.getDeviceAttributeLongLong('excitation_frequency');
```
Comment (tfcollins) [inline +adi/+AD2S1210/Rx.m:120]: This should be a write and not a read. MATLAB translates states at initialization based on MATLAB's pre-connection state

**Review by tfcollins — CHANGES_REQUESTED** [review]
Need some updates related to the get methods

File: +adi/+AD2S1210/Rx.m (hunk 3 above)
Comment (apelete) [reply +adi/+AD2S1210/Rx.m:111]: Thanks for noticing, fixed.

File: +adi/+AD2S1210/Rx.m (hunk 4 above)
Comment (apelete) [reply +adi/+AD2S1210/Rx.m:39]: Done.

File: +adi/+AD2S1210/Rx.m (hunk 5 above)
Comment (apelete) [reply +adi/+AD2S1210/Rx.m:120]: Reworded "Read" into "Write" in that comment to better reflect what initialization method is doing here.

**Comment by apelete** [comment]
> Need some updates related to the get methods

Updated get methods for Dependent properties.

File: +adi/+AD2S1210/Rx.m [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+        VelocityScale
+    end
+
+    properties
+        % ExcitationFrequency Excitation Frequency
+        %   Resolver excitation frequency in Hertz.
+        ExcitationFrequency = 0;
+    end
+
+    properties (Hidden)
```
Comment (ribdp) [inline +adi/+AD2S1210/Rx.m:43]: This property is not being used - can be removed.

Comment (apelete) [reply +adi/+AD2S1210/Rx.m:43]: Attribute was removed from the driver indeed ; removing it here too to stay in sync.

Comment (ribdp) [reply +adi/+AD2S1210/Rx.m:37]: Ah, I wasn't aware ExcitationFrequency had been removed from the driver. I was actually referring to the FramesCount property.

Comment (ribdp) [reply +adi/+AD2S1210/Rx.m:37]: It looks like the attribute is still there in the driver? https://github.com/analogdevicesinc/linux/blob/71da83b0b094a92ed4c2b8c21503638b3557ac56/drivers/staging/iio/resolver/ad2s1210.c#L609

Comment (apelete) [reply +adi/+AD2S1210/Rx.m:37]: > Ah, I wasn't aware ExcitationFrequency had been removed from the driver. I was actually referring to the FramesCount property.

FramesCount property seems to be mandatory; here is the error message I get whent trying to remove it :

```
Abstract classes cannot be instantiated. 
Class 'adi.AD2S1210.Rx' inherits abstract methods or properties but does not implement them.
See the list of methods and properties that 'adi.AD2S1210.Rx' must implement if you do not intend the class to be abstract.

Abstract properties for class adi.AD2S1210.Rx:
    SamplesPerFrame	% defined in matlabshared.libiio.base
```

> It looks like the attribute is still there in the driver? https://github.com/analogdevicesinc/linux/blob/71da83b0b094a92ed4c2b8c21503638b3557ac56/drivers/staging/iio/resolver/ad2s1210.c#L609

There's actually an upstream proposal to convert it to a channel attribute (thus warranting the need to remove it from the Matlab code for now):
https://lore.kernel.org/all/20230929-ad2s1210-mainline-v3-17-fa4364281745@baylibre.com/

**Comment by apelete** [comment]
@ribdp @tfcollins AFAIK requested changes were implemented so far : any update on this ?

**Comment by apelete** [comment]
> sysobjs.json

Thanks for the quick reply : latest commit updates sysobjs.json as per your instructions.

**Comment by ribdp** [comment]
Confirmed documentation gets generated properly: https://ribdp.github.io/PrecisionToolboxForked/master/sysobjects/adi.AD2S1210.Rx/

Looks good to me.

**Review by tfcollins — APPROVED** [review]

**Review by ribdp — APPROVED** [review]

**Comment by apelete** [comment]
@ribdp @tfcollins I rebased and signed all commits with GPG to get them verified, can you please merge this pull request into main branch ?
I am not able to do so through GitHub webui, getting the following error when clicking on "Rebase and merge" button :
```
Merge attempt failed

Base branch requires signed commits. Rebase merges cannot be automatically signed by GitHub
```

**Comment by apelete** [comment]
@ribdp @tfcollins There may be a merge issue when signing the commits with GPG according to documentation (https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#signature-verification-for-rebase-and-merge) : 
> When using the Rebase and Merge option on a pull request, it's important to note that the commits in the head branch are added to the base branch without commit signature verification. When you use this option, GitHub creates a modified commit, using the data and content of the original commit. This means that GitHub didn't truly create this commit, and can't therefore sign it as a generic system user. GitHub doesn't have access to the committer's private signing keys, so it can't sign the commit on the user's behalf.
> 
> A workaround for this is to rebase and merge locally, and then push the changes to the pull request's base branch.

Do you want me to squash and merge all commits before you push the resulting single commit into the main branch to work around this issue ?

**Comment by ribdp** [comment]
Hi @apelete I do see another issue with the merge check, although upon looking into it it's not related to this PR's changes. So if you could just resolve the merge conflicts, I'd merge and close both PRs. Squashing might not be necessary now, but if it is I'll do that as well.

**Comment by apelete** [comment]
> Hi @apelete I do see another issue with the merge check, although upon looking into it it's not related to this PR's changes. So if you could just resolve the merge conflicts, I'd merge and close both PRs. Squashing might not be necessary now, but if it is I'll do that as well.

@ribdp I rebased on main branch again and solved the conflicts, should be good for merging now.

**Pull request merged**